src/main/scala/nl/knaw/dans/easy/multideposit/Action.scala

```diff
@@ -85,101 +92,92 @@ trait Action extends DebugEnhancedLogging {
+            case Failure(e2) => reportFailure(CompositeException(es ++ List(e2)))
+          }
+        case NonFatal(e1) =>
+          rollback() match {
+            case Success(_) => reportFailure(e1)
+            case Failure(CompositeException(es2)) => reportFailure(CompositeException(List(e1) ++ es2))
+            case Failure(e2) => reportFailure(CompositeException(List(e1, e2)))
+          }
       }
-    } yield ()
+    } yield t
   }
 
   private def generateReport(header: String = "", throwable: Throwable, footer: String = ""): String = {
-    val report = throwable match {
-      case ActionException(row, msg, _) => s" - row $row: $msg"
-      case CompositeException(throwables) =>
-        throwables.toSeq
-          .collect {
-            case ActionException(row, msg, _) => s" - row $row: $msg"
-            case NonFatal(e) => s" - unexpected error: ${e.getMessage}"
-          }
-          .mkString("\n")
-      case e => s" - unexpected error: ${e.getMessage}"
+
+    @tailrec
+    def report(es: List[Throwable], rpt: List[String] = Nil): List[String] = {
+      es match {
+        case Nil => rpt
+        case ActionException(row, msg, _) :: xs => report(xs, s" - row $row: $msg" :: rpt)
+        case CompositeException(ths) :: xs => report(ths.toList ::: xs, rpt)
+        case NonFatal(ex) :: xs => report(xs, s" - unexpected error: ${ex.getMessage}" :: rpt)
+      }
     }
 
-    header.toOption.map(_ + "\n").getOrElse("") + report + footer.toOption.map("\n" + _).getOrElse("")
+    header.toOption.fold("")(_ + "\n") +
+      report(List(throwable)).reverse.mkString("\n") +
+      footer.toOption.fold("")("\n" + _)
   }
 
   /**
-   * Compose this action with `other` into a new action such that it runs both actions in the correct order.
-   * [[checkPreconditions]] and [[execute]] are run in order, [[rollback]] is run in reversed order.
+   * Sequentially composes two `Action`s by running their `execute` methods one after the other.
+   * In this, the input of `other` is the output of this `Action`.
    *
-   * @param other the second action to be run
-   * @return a composed action
+   * @param other the `Action` to combine this `Action` with
+   * @tparam S the output type of the second `Action`
+   * @return an `Action` that composes these two actions sequentially
    */
-  def compose(other: Action): ComposedAction = {
-    other match {
-      case composedAction: Action#ComposedAction => ComposedAction(this :: composedAction.actions)
-      case action => ComposedAction(this :: action :: Nil)
-    }
-  }
+  def combine[S](other: Action[T, S]): Action[A, S] = new Action[A, S] {
+    private var pastSelf = false
 
-  final case class ComposedAction(actions: List[Action]) extends Action {
-    private lazy val executedActions = mutable.Stack[Action]()
+    override def run(a: A): Try[S] = {
+      pastSelf = false
+      super.run(a)
+    }
```

Comment:
this makes sure the `pastSelf` is set to false. After you run the composed action once, the `pastSelf` is set to `true`. When a new run starts, you need to reset `pastSelf` to `false`.